src/js/modal.js

```diff
@@ -0,0 +1,156 @@
+export function createModal(challenge) {
+
+    const container = document.createElement('div');
+    container.classList.add('modal-container');
+    document.body.appendChild(container);
+
+    renderContent(step1(challenge));
+    
+    return container;
+}
+
+export function removeModal() {
+    document.querySelector('.modal-container').remove();
+}
+
+function renderContent(content) {
+    const container = document.querySelector('.modal-container'); 
+    container.innerHTML = '';
+    container.appendChild(content);
+}
+
+function step1(challenge) {
+
+    const form = document.createElement('form');
+    form.classList.add('content');
+
+    const h1 = document.createElement('h1');
+    h1.textContent = `Book room ${challenge.title} (step 1)`;
+    form.appendChild(h1);
+
+    const p = document.createElement('p');
+    p.textContent = 'What date would you like to come?';
+    form.appendChild(p);
+
+    const label = document.createElement('label');
+    label.textContent = 'Date';
+    label.for = 'input[type=date]';
+    form.appendChild(label);
+
+    const inputDate = document.createElement('input');
+    inputDate.type = 'date';
+    form.appendChild(inputDate);
+
+    const button = document.createElement('button');
+    button.type = 'submit';
+    button.classList.add('button', 'primary');
+    button.name = 'search-times';
+    button.textContent = 'Search available times';
+    form.appendChild(button);
+
+    form.addEventListener('submit', async e => {
+        e.preventDefault();
+
+        const times = await loadTimes(); //Martas function här
```

Comment:
Correct me if I'm wrong, but I don't hink the `loadTimes()` function exists at this point? It's perfectly fine to develop one aspect of a program without all other pieces being there yet, but you should usually handle it not by calling a function that does not exist (which will raise errors) but by calling a placeholder function that does nothing (or something else than what it will eventually do).

I would either just put a global `loadTimes()` function somewhere, or I would build the modal code such that the function to call can be injected one way or another.

In my example, I achieved this using the "API adapter", but it's also possible to just pass a function reference when creating the modal, e.g. `createModal(challenge, functionThatDoesNothing)`.

This way, the code in this module is complete and working, but leaves it to the user of the code to supply external dependencies. That's what we call "dependency injection".

Reply:
Correct! As you said this was a problem when someone wanted to test the code and got an instant error. It doesn't seem like the most efficient way of testing when you have to deactivate code (or make your own placeholder) to make it work :)

Great tip! I would probably do dependency injection if I would redo this. Seems like the more secure solution and perhaps has better seperation of concern? :)